Support SNMPv3 and asyncio in snmp sensor

mtdcr opened the pull request:
Instantiate pysnmp objects once during setup and reuse them. Requires asyncio, because SnmpEngine objects can't be shared between threads. This speeds up repeated queries significantly and simplifies support for different authentication types.

Related to #6973 and #11370.

For SNMPv3, the following new options were modified or added:

* **version**
  Changed to accept version '3'.
* **username**
   **`<string>`**
* **auth_key**
   `<string>`
* **auth_protocol**
  `<none|hmac-md5|hmac-sha|hmac128-sha224|hmac192-sha256|hmac256-sha384|hmac384-sha512>`
* **priv_key**
   `<string>`
* **priv_protocol**
   `<none|des|3des-ede|aes-cfb-128|aes-cfb-192|aes-cfb-256>`

Note:
* auth_protocol defaults to `none`.
* priv_protocol defaults to `none`.
* Setting no auth_key forces auth_protocol to `none`.
* Setting no priv_key forces priv_protocol to `none`.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5484

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

No new files or dependencies were introduced.

mtdcr committed:
snmp sensor: Add asyncio support and reuse SnmpEngine object

mtdcr committed:
snmp sensor: Support protocol v3

homeassistant commented:
Hi @mtdcr, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#14753).

Once you do that we will be able to review and accept this pull request.

Thanks!

mtdcr commented:
TOXENV=pylint fails on Travis CI:
```
E:161,40: No value for argument 'transportTarget' in function call (no-value-for-parameter)
E:161,40: No value for argument 'contextData' in function call (no-value-for-parameter)
```
I'd say that's not true. And it doesn't occur when I run the tests locally.

mtdcr commented:
I managed to reproduce the warning locally and successively replaced the code in question with functionally equivalent code accepted by pylint.

dgomes commented:
We have adopted PEP 492, please update accordingly, eg:

```
@asyncio.coroutine
def function():`
```
to 

```
async def function
```

`yield from` becomes `await`

pvizeli commented on homeassistant/components/sensor/snmp.py:
```diff
@@ -50,13 +77,27 @@
+    vol.Optional(CONF_AUTH_KEY): cv.string,
+    vol.Optional(CONF_AUTH_PROTOCOL, default=DEFAULT_AUTH_PROTOCOL):
+        vol.In(AUTH_PROTOCOLS),
+    vol.Optional(CONF_PRIV_KEY): cv.string,
```
is that a file, use cv.isfile

mtdcr replied:
No, it's a string.

mtdcr commented:
@dgomes: I've updated my branch.

pvizeli commented on homeassistant/components/sensor/snmp.py:
```diff
@@ -66,27 +105,69 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+            UsmUserData(
+                username,
+                authKey=authkey or None,
+                privKey=privkey or None,
```
Need not set to None

mtdcr replied:
This replaces en empty string with None, which makes a difference in UsmUserData. Are you implying that authkey=="" is guaranteed not to appear in this codepath?

mtdcr commented:
@pvizeli: Thanks for your suggestions! I've updated my branch.

fabaff commented on homeassistant/components/sensor/snmp.py:
```diff
@@ -50,14 +76,22 @@
+    vol.Optional(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_AUTH_KEY): cv.string,
+    vol.Optional(CONF_AUTH_PROTOCOL, default=DEFAULT_AUTH_PROTOCOL):
+        vol.In(MAP_AUTH_PROTOCOLS.keys()),
```
`.keys()` is not needed, it's the default.

fabaff commented on homeassistant/components/sensor/snmp.py:
```diff
@@ -50,14 +76,22 @@
+        vol.In(MAP_AUTH_PROTOCOLS.keys()),
+    vol.Optional(CONF_PRIV_KEY): cv.string,
+    vol.Optional(CONF_PRIV_PROTOCOL, default=DEFAULT_PRIV_PROTOCOL):
+        vol.In(MAP_PRIV_PROTOCOLS.keys()),
```
Dito

mtdcr commented:
I rebased this branch on dev again and removed keys(), even though using keys() was suggested to use before. Contributing to this project is fun, really.

MartinHjelmare reviewed (approved):
Looks good to me.
@pvizeli @fabaff ?

fabaff committed:
Merge branch 'dev' into snmpv3-asyncio-sensor

fabaff committed:
Fix lint issue
